File: ts/src/core/identifier-ref.ts
/**
 * YANG identifier-ref: an identifier with an optional module prefix
 * (RFC 7950 prefix:identifier). Built once at parse time — transformers
 * must not re-split qname strings.
 */
export type YangIdentifierRef = {
  /** Import / module prefix when present; undefined for an unprefixed name. */
  prefix?: string;
  /** Local identifier. */
  name: string;
};

export function formatIdentifierRef(ref: YangIdentifierRef): string {
  return ref.prefix ? `${ref.prefix}:${ref.name}` : ref.name;
}

export function identifierRef(name: string, prefix?: string | undefined): YangIdentifierRef {
  return prefix ? { prefix, name } : { name };
}

/**
 * Convert a single opaque atom (if-feature token, JSON Schema base string,
 * RFC 7951 identityref instance value) into an identifier-ref.
 * Not for YANG token streams — those use ``consume_identifier_ref``.
 */
export function parseIdentifierRefAtom(atom: string): YangIdentifierRef {
  const idx = atom.indexOf(":");
  if (idx <= 0 || idx >= atom.length - 1) {
    return { name: atom };
  }
  return { prefix: atom.slice(0, idx), name: atom.slice(idx + 1) };
}

/**
 * Parse an absolute schema-node path (``/prefix:a/prefix:b``) into segment refs.
 * Used when a path arrives as a YANG string argument (augment, rfc8791).
 */
export function parseAbsoluteSchemaPath(path: string): YangIdentifierRef[] {
  const raw = path.replace(/^["']|["']$/g, "");
  if (!raw.startsWith("/")) {
    throw new Error(`Schema path must be absolute, got '${path}'`);
  }
  const parts = raw
    .slice(1)
    .split("/")
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
  if (parts.length === 0) {
    throw new Error(`Empty schema path: '${path}'`);
  }
  return parts.map((seg) => {
    const ref = parseIdentifierRefAtom(seg);
    if (!ref.prefix) {
      throw new Error(`Invalid schema path segment '${seg}': expected 'prefix:identifier'`);
    }
    return ref;
  });
}

/** Normalize JSON/legacy string-or-object bases into identifier-refs. */
export function coerceIdentifierRef(value: unknown): YangIdentifierRef | undefined {
  if (typeof value === "string" && value.length > 0) {
    return parseIdentifierRefAtom(value);
  }
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const o = value as { name?: unknown; prefix?: unknown };
    if (typeof o.name === "string" && o.name.length > 0) {
      return typeof o.prefix === "string" && o.prefix.length > 0
        ? { prefix: o.prefix, name: o.name }
        : { name: o.name };
    }
  }
  return undefined;
}

File: src/xyang/identifier_ref.py
"""
YANG identifier-ref: an identifier with an optional module prefix
(RFC 7950 prefix:identifier). Built once at parse time — transformers
must not re-split qname strings.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Optional


@dataclass(frozen=True)
class YangIdentifierRef:
    """Local identifier with optional import / module prefix."""

    name: str
    prefix: Optional[str] = None


def format_identifier_ref(ref: YangIdentifierRef) -> str:
    return f"{ref.prefix}:{ref.name}" if ref.prefix else ref.name


def identifier_ref(name: str, prefix: Optional[str] = None) -> YangIdentifierRef:
    return YangIdentifierRef(name=name, prefix=prefix) if prefix else YangIdentifierRef(name=name)


def parse_identifier_ref_atom(atom: str) -> YangIdentifierRef:
    """Convert one opaque atom (if-feature token, JSON Schema base string, RFC 7951 value)."""
    idx = atom.find(":")
    if idx <= 0 or idx >= len(atom) - 1:
        return YangIdentifierRef(name=atom)
    return YangIdentifierRef(prefix=atom[:idx], name=atom[idx + 1 :])


def parse_absolute_schema_path(path: str) -> list[YangIdentifierRef]:
    """Parse absolute schema-node path (``/prefix:a/prefix:b``) into segment refs."""
    raw = path.strip().strip('"').strip("'")
    if not raw.startswith("/"):
        raise ValueError(f"Schema path must be absolute, got {path!r}")
    parts = [p.strip() for p in raw[1:].split("/") if p.strip()]
    if not parts:
        raise ValueError(f"Empty schema path: {path!r}")
    out: list[YangIdentifierRef] = []
    for seg in parts:
        ref = parse_identifier_ref_atom(seg)
        if not ref.prefix:
            raise ValueError(
                f"Invalid schema path segment {seg!r}: expected 'prefix:identifier'"
            )
        out.append(ref)
    return out


def coerce_identifier_ref(value: Any) -> Optional[YangIdentifierRef]:
    """Normalize JSON/legacy string-or-object bases into identifier-refs."""
    if isinstance(value, str) and value:
        return parse_identifier_ref_atom(value)
    if isinstance(value, dict):
        name = value.get("name")
        if isinstance(name, str) and name:
            prefix = value.get("prefix")
            if isinstance(prefix, str) and prefix:
                return YangIdentifierRef(prefix=prefix, name=name)
            return YangIdentifierRef(name=name)
    if isinstance(value, YangIdentifierRef):
        return value
    return None
